numpy/core/src/umath/loops.c.src

```diff
@@ -2618,41 +2618,64 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
 NPY_NO_EXPORT void
 OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
 {
-#if defined(NPY_PY3K)
     PyObject *zero = PyLong_FromLong(0);
+
     UNARY_LOOP {
         PyObject *in1 = *(PyObject **)ip1;
         PyObject **out = (PyObject **)op1;
+        PyObject *ret = NULL;
         int v;
-        PyObject *ret;
-        if (PyObject_Cmp(in1 ? in1 : Py_None, zero, &v) == -1) {
-            return;
+
+        if (in1 == NULL) {
+            in1 = Py_None;
         }
-        ret = PyLong_FromLong(v);
-        if (PyErr_Occurred()) {
-            Py_DECREF(zero);
-            return;
+
+        v = PyObject_RichCompareBool(in1, zero, Py_LT);
+        if (v == 1) {
+            ret = PyLong_FromLong(-1);
+            goto done;
+        } else if (v == -1) {
+            goto fail;
         }
-        Py_XDECREF(*out);
-        *out = ret;
-    }
-    Py_DECREF(zero);
-#else
-    PyObject *zero = PyInt_FromLong(0);
-    UNARY_LOOP {
-        PyObject *in1 = *(PyObject **)ip1;
-        PyObject **out = (PyObject **)op1;
-        PyObject *ret = PyInt_FromLong(
-                            PyObject_Compare(in1 ? in1 : Py_None, zero));
-        if (PyErr_Occurred()) {
-            Py_DECREF(zero);
-            return;
+
+        v = PyObject_RichCompareBool(in1, zero, Py_GT);
+        if (v == 1) {
+            ret = PyLong_FromLong(1);
+            goto done;
+        } else if (v == -1) {
+            goto fail;
         }
-        Py_XDECREF(*out);
-        *out = ret;
+
+        v = PyObject_RichCompareBool(in1, zero, Py_EQ);
+        if (v == 1) {
+            ret = PyLong_FromLong(0);
+            goto done;
+        } else if (v == -1) {
+            goto fail;
+        }
+
+        if (v == 0) {
+            /* In this case, we are dealing with NaN */
+            PyErr_SetString(PyExc_TypeError, "unorderable types for comparison");
+            goto fail;
+        }
+
+        fail:
+            if (PyErr_Occurred()) {
+                Py_DECREF(zero);
+                return;
+            }
+            /* In case of error in comparison */
+            if (v == -1) {
```

Comment:
For this, should this be
```c
if (v == 1) {
        ret = PyLong_FromLong(0);
        goto done;
}
else if (v == -1) {
        break;                                                                                                                                                  
}
done: 
        Py_XDECREF(*out);
        *out = ret;

```
And use `PyDECREF(zero);` outside the `UNARY LOOP` in case of failure ?